clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -90,20 +89,259 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    // 0.
+    if (dstType->isBooleanType())
+      return emitConversionToBool(src, srcType, cgf.getLoc(loc));
+
+    mlir::Type mlirDstType = cgf.convertType(dstType);
+
+    if (srcType->isHalfType() &&
+        !cgf.getContext().getLangOpts().NativeHalfType) {
+      // Cast to FP using the intrinsic if the half type itself isn't supported.
+      if (mlir::isa<cir::CIRFPTypeInterface>(mlirDstType)) {
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics())
+          cgf.getCIRGenModule().errorNYI(loc,
+                                         "cast via llvm.convert.from.fp16");
+      } else {
+        // Cast to other types through float, using either the intrinsic or
+        // FPExt, depending on whether the half type itself is supported (as
+        // opposed to operations on half, available with NativeHalfType).
+        if (cgf.getContext().getTargetInfo().useFP16ConversionIntrinsics()) {
+          cgf.getCIRGenModule().errorNYI(loc,
+                                         "cast via llvm.convert.from.fp16");
+          // FIXME(cir): For now lets pretend we shouldn't use the conversion
+          // intrinsics and insert a cast here unconditionally.
+        }
+        src = builder.createCast(cgf.getLoc(loc), cir::CastKind::floating, src,
+                                 cgf.FloatTy);
+        srcType = cgf.getContext().FloatTy;
+        mlirSrcType = cgf.FloatTy;
+      }
+    }
+
+    // TODO(cir): LLVM codegen ignore conversions like int -> uint,
+    // is there anything to be done for CIR here?
+    if (mlirSrcType == mlirDstType) {
+      if (opts.emitImplicitIntegerSignChangeChecks)
+        cgf.getCIRGenModule().errorNYI(loc,
+                                       "implicit integer sign change checks");
+      return src;
+    }
+
+    // Handle pointer conversions next: pointers can only be converted to/from
+    // other pointers and integers. Check for pointer types in terms of LLVM, as
+    // some native types (like Obj-C id) may map to a pointer type.
+    if (auto dstPT = dyn_cast<cir::PointerType>(mlirDstType)) {
+      cgf.getCIRGenModule().errorNYI(loc, "pointer casts");
+    }
+
+    if (isa<cir::PointerType>(mlirSrcType)) {
+      // Must be an ptr to int cast.
+      assert(isa<cir::IntType>(mlirDstType) && "not ptr->int?");
+      return builder.createPtrToInt(src, mlirDstType);
+    }
+
+    // A scalar can be splatted to an extended vector of the same element type
+    if (dstType->isExtVectorType() && !srcType->isVectorType()) {
+      // Sema should add casts to make sure that the source expression's type
+      // is the same as the vector's element type (sans qualifiers)
+      assert(dstType->castAs<ExtVectorType>()->getElementType().getTypePtr() ==
+                 srcType.getTypePtr() &&
+             "Splatted expr doesn't match with vector element type?");
+
+      llvm_unreachable("not implemented");
```

Comment:
Any reason this is a `unreachable` instead of a NYI?  Also, `unreachable` isnt' really the tool for `not yet implemented`, we typically want to use `assert` for reasons that the lead maintainer explained to me in the past but I've failed to internalize.